File: flutter/multando_sdk/lib/src/services/blockchain_service.dart
import '../core/http_client.dart';
import '../models/blockchain.dart';

/// Service for blockchain token operations: balance, staking, transactions.
class BlockchainService {
  BlockchainService({required MultandoHttpClient httpClient})
      : _http = httpClient;

  final MultandoHttpClient _http;

  /// Get the current user's token balance.
  Future<TokenBalance> getBalance() async {
    final response =
        await _http.get<Map<String, dynamic>>('/blockchain/balance');
    return TokenBalance.fromJson(response.data!);
  }

  /// Stake tokens.
  Future<StakingInfo> stake(StakeRequest request) async {
    final response = await _http.post<Map<String, dynamic>>(
      '/blockchain/stake',
      data: request.toJson(),
    );
    return StakingInfo.fromJson(response.data!);
  }

  /// Unstake tokens.
  Future<StakingInfo> unstake(UnstakeRequest request) async {
    final response = await _http.post<Map<String, dynamic>>(
      '/blockchain/unstake',
      data: request.toJson(),
    );
    return StakingInfo.fromJson(response.data!);
  }

  /// Get current staking information.
  Future<StakingInfo> stakingInfo() async {
    final response =
        await _http.get<Map<String, dynamic>>('/blockchain/staking-info');
    return StakingInfo.fromJson(response.data!);
  }

  /// List token transactions with pagination.
  Future<List<TokenTransaction>> transactions({
    int page = 1,
    int pageSize = 20,
  }) async {
    final response = await _http.get<List<dynamic>>(
      '/blockchain/transactions',
      queryParameters: {
        'page': page,
        'page_size': pageSize,
      },
    );
    return response.data!
        .cast<Map<String, dynamic>>()
        .map(TokenTransaction.fromJson)
        .toList();
  }

  /// Claim pending staking rewards.
  Future<ClaimRewardsResponse> claimRewards() async {
    final response = await _http.post<Map<String, dynamic>>(
      '/blockchain/claim-rewards',
    );
    return ClaimRewardsResponse.fromJson(response.data!);
  }
}

File: react-native/src/services/blockchainService.ts
import { AxiosInstance } from 'axios';
import { Logger } from '../core/logger';
import {
  TokenBalance,
  StakeRequest,
  StakingInfo,
  TokenTransaction,
  ClaimRewardsResponse,
} from '../models/blockchain';

export class BlockchainService {
  private http: AxiosInstance;
  private logger: Logger;

  constructor(http: AxiosInstance, logger: Logger) {
    this.http = http;
    this.logger = logger;
  }

  async getBalance(): Promise<TokenBalance> {
    const response = await this.http.get<TokenBalance>('/blockchain/balance');
    return response.data;
  }

  async stake(request: StakeRequest): Promise<StakingInfo> {
    this.logger.info('Staking tokens', { amount: request.amount });
    const response = await this.http.post<StakingInfo>(
      '/blockchain/stake',
      request,
    );
    return response.data;
  }

  async unstake(request: StakeRequest): Promise<StakingInfo> {
    this.logger.info('Unstaking tokens', { amount: request.amount });
    const response = await this.http.post<StakingInfo>(
      '/blockchain/unstake',
      request,
    );
    return response.data;
  }

  async getStakingInfo(): Promise<StakingInfo> {
    const response = await this.http.get<StakingInfo>(
      '/blockchain/staking-info',
    );
    return response.data;
  }

  async getTransactions(): Promise<TokenTransaction[]> {
    const response = await this.http.get<TokenTransaction[]>(
      '/blockchain/transactions',
    );
    return response.data;
  }

  async claimRewards(): Promise<ClaimRewardsResponse> {
    this.logger.info('Claiming rewards');
    const response = await this.http.post<ClaimRewardsResponse>(
      '/blockchain/claim-rewards',
    );
    return response.data;
  }
}
